alexandria/storage/src/list.cairo

```diff
@@ -0,0 +1,215 @@
+use hash::LegacyHash;
+use integer::U32DivRem;
+use option::OptionTrait;
+use starknet::{storage_read_syscall, storage_write_syscall, SyscallResult, SyscallResultTrait};
+use starknet::storage_access::{
+    StorageAccess, StorageBaseAddress, storage_address_to_felt252, storage_address_from_base,
+    storage_address_from_base_and_offset, storage_base_address_from_felt252
+};
+use traits::{Default, DivRem, IndexView, Into, TryInto};
+
+const POW2_8: u32 = 256; // 2^8
+
+#[derive(Drop)]
+struct List<T> {
+    address_domain: u32,
+    base: StorageBaseAddress,
+    len: u32, // number of elements in array
+    storage_size: u8
+}
+
+trait ListTrait<T> {
+    fn len(self: @List<T>) -> u32;
+    fn is_empty(self: @List<T>) -> bool;
+    fn append(ref self: List<T>, value: T) -> u32;
+    fn get(self: @List<T>, index: u32) -> Option<T>;
+    fn set(ref self: List<T>, index: u32, value: T);
+    fn pop_front(ref self: List<T>) -> Option<T>;
+}
+
+// when writing elements in storage, we need to know how many storage slots
+// they take up to properly calculate the offset into a storage segment
+// that's what this trait provides
+// it's very similar to the `size_internal` function in StorageAccess trait
+// with one crutial difference - this function does not need a T element
+// to return the size of the T element!
+trait StorageSize<T> {
+    fn storage_size() -> u8;
+}
+
+// any type that can Into to u128 has a size of 1
+impl StorageSizeU128<T, impl TInto: Into<T, u128>> of StorageSize<T> {
+    fn storage_size() -> u8 {
+        1
+    }
+}
+
+// any type that can Into to felt252 has a size of 1
+impl StorageSizeFelt252<T, impl TInto: Into<T, felt252>> of StorageSize<T> {
+    fn storage_size() -> u8 {
+        1
+    }
+}
+
+// any type that can Into to u258 has a size of 2
```

Comment:
Hmm, that's true. In fact `U128IntoU256` is already present in corelib.

Not sure how Cairo handles that because there's also `TIntoT` and `U128IntoFelt252`. I presume there's some kind of priority list when resolving these, but no idea how that works or which `StorageSize` implementation would be used here for `u128` 😬

Reply:
So what happens in that case is the compiler fails with an `Trait `core::starknet::storage_access::StorageAccess::<<missing>>` has multiple implementations` error. 

I think the best course of action here is to keep `StorageSizeFelt252` (that would cover most of corelib IMO) and change `StorageSizeU256` to match only `u256`.

Reply:
I'm sure we could have a way to automatically get this at compile time similarly to struct.SIZE in cairo 0 but anyway lgtm